README
#------------------------------------------------------------------------------------
#  SusyNtuple
#
#  A RootCore package for common UCI SUSY analysis
#------------------------------------------------------------------------------------


# For useful information on using these packages, please consult the TWiki:
https://twiki.cern.ch/twiki/bin/viewauth/AtlasProtected/UCISusyNtuples

# Link to code in SVN browser
https://svnweb.cern.ch/trac/atlasinst/browser/Institutes/UCIrvine/SUSYAnalysis/SusyNtuple

# To checkout the package trunk
svn co svn+ssh://svn.cern.ch/reps/atlasinst/Institutes/UCIrvine/SUSYAnalysis/SusyNtuple/trunk SusyNtuple

# Or the current recommended tag (get the latest tag from the TWiki)
svn co svn+ssh://svn.cern.ch/reps/atlasinst/Institutes/UCIrvine/SUSYAnalysis/SusyNtuple/tags/SusyNtuple-XX-YY-ZZ SusyNtuple

# To make a tag, use the script/makeTag.sh script (only for developers)
./scripts/makeTag.sh SusyNtuple-XX-YY-ZZ "Some message"



# Note that if you are not planning on producing SusyNt files, you do not need 
# to check out the SusyCommon, MultiLep, or SUSYTools packages. Instead, you 
# can use this install script which will checkout the minimal RootCore package
# content necessary to read and analyze SusyNt. It will check out a few classes
# from SUSYTools rather than the entire package with all of its dependencies.
source SusyNtuple/scripts/installMinimalSUSYTools.sh

# If you've already done the minimal installation above but want to update
# the SUSYTools/data content (e.g., to pick up updated cross sections),
# then use the following script.
source SusyNtuple/scripts/refreshSUSYToolsData.sh

# If you plan to produce SusyNt or just want to have the whole machinery 
# available to look at, then check out the SusyCommon package and use the 
# install instructions from there. In that case, this package will be checked 
# out automatically.

# for examples of how to utilize this package, you can see the example cutflow 
# codes below, or see my personal analysis code here:
https://svnweb.cern.ch/trac/atlasinst/browser/Institutes/UCIrvine/sfarrell/SusyAna


#------------------------------------------------------------------------------------
# Examples on how to run
#------------------------------------------------------------------------------------

# We have include some example codes that demonstrate the usage of SusyNt and 
# can reproduce the 2L and 3L cutflows.
# It is recommended that you write your own analysis loopers which inherit from SusyNtAna
# (or from the cutflow classes)

# Currently there are 3 executables.  They all have the same command line options.
# Specify SusyNt files either by specific file path, a fileList, or an input directory

# Executable for simple SusyNt testing
SusyNtTest

# Executables 2L and 3L cutflows
Susy2LepCF
Susy3LepCF

# If you run with option '-h' you will get the list of command line options
  Options:
  -n number of events to process
     defaults: -1 (all events)
  -k number of events to skip
     defaults: 0
  -d debug printout level
     defaults: 0 (quiet) 
  -F name of single input file
     defaults: ''
  -f name of input filelist
     defaults: ''
  -D name of input file dir
     defaults: ''
  -s sample name, for naming files
     defaults: ntuple sample name
  -h print this help



#------------------------------------------------------------------------------------
# Overview of the package
#------------------------------------------------------------------------------------

SusyDefs
  - typedefs, constants, globals, etc.

SusyNt
  - Definitions for the SusyNt classes: particles, jets, event variables, etc.

SusyNtObject
  - Manages SusyNt read/write of branches

SusyNtTools
  - Class with low-level functions for applying selections on SusyNt files

SusyNtAna
  - Analysis looper class, inherits from TSelector, used to process the SusyNt
  - Also inherits from SusyNtTools and allows to easily select all baseline and signal objects

DilTrigLogic
  - Class which implements the current 2-lepton trigger logic

TrilTrigLogic
  - Class which implements the current 3-lepton trigger logic

WhTruthExtractor
  - A class to extract from d3pd the truth information for simplifiedModel_wA_noslep_WH samples
  - This should maybe be moved into SusyCommon

SusyNtTruthAna
  - A looper class for analyzing only the truth particles from SusyNt


ChangeLog
# vim: set noexpandtab:

2014-10-30 Serhan Mete <amete@cern.ch>
	* Added SV1plusIP3D b-tag weight for jets in xAOD DC14
	modified   Root/SusyNt.cxx
	modified   SusyNtuple/SusyNt.h

2014-10-01 Serhan Mete <amete@cern.ch>
	modified:   Root/MCWeighter.cxx
	modified:   SusyNtuple/MCWeighter.h
	* Drop buildSumwMap from constructor, user needs to call it,
	this gives the freedom to invoke setBinCounter before buildingSumwMap,
	which is necessary for Simplified Models
	* Set defaultLabelBinCounter to "Initial" from "SusyProp Veto"

2014-04-02 Steve
	* Added deprecation warnings to SusyNtTools methods: buildSumwMap and getEventWeight
		* Users should update to usage of the MCWeighter class,
		  as demonstrated in the Susy3LepCutflow class

2014-03-26 Steve
	* Moved JVF sys enums to SusyNtSys
		* Kinda goes against my thinking that this enum should
		  only be for variations that are saved in the ntuple,
		  but otherwise I'd have to add an additional sys enum
		  argument to the selectObjects methods, so this is
		  cleaner to the user
	* Added JVF systematic to SusyNtTools
		* JVFUncertaintyTool instantiated in SusyNtTools
		* Controlled via usual object syst enum, SusyNtSys,
		  when calling selectObjects method
		* I had to make some of the jet counting methods
		  (and getMetRel) non-static to accommodate the usage
		  of the JVFUncertaintyTool

2014-03-25 Steve
	* Added JVF sys enums to SusyDefs.h

2014-03-21 Steve
	* Added JVFUncertaintyTool to installMinimalSUSYTools script,
	  necessary for applying the JVF systematic

2014-03-08 Steve
	* Updated MCWeighter
		* Now accepts a TTree for building the sumw map,
		  and has dedicated methods for TTree and TChain
		* TTree arg to constructor no longer default
	* Updated usage of MCWeighter in Susy3LepCutflow
		* Now instantiated and initialized in Init method rather than main.
		  It cannot be initialized in Begin because the tree pointer isn't set.
		  	* This is just one way to initialize the MCWeighter.
			* NOTE: This is explicitly incompatible with PROOF. But then
			  again, I don't think the rest of our code is PROOF-compatible
			  anyway.
	* MCWeighter TODO
		* Update usage in Susy2LepCutflow
		* Change SusyNtAna and SusyNtTools methods to use 
		  MCWeighter behind the scenes.
	* Added second version of template dumpObjects method to SusyNtTools,
	  a version for vectors of objects as well as the original one for
	  vectors of object pointers.

2014-03-03 Steve
	* Code improvements for MCWeighter

2014-03-02 Steve
	* Added full functionality to MCWeighter class
		* getMCWeight, getSumw, getCrossSection, getXsecTimesEff, getPileupWeight
		* Also added methods and enums for controlling the sumw and xsec methods
		* Currently copies the functionality in SusyNtTools. I haven't removed
		  the old approach yet.
	* Made BTagCalib a static pointer in SusyNtTools
		* This is to reduce the number of instantiations of the class
		  and open root files polluting the ROOT space when a user
		  uses many instantiations of SusyNtTools (like I do).
		  If this causes problems for users, I can change it back.
	* Updated Susy3LepCutflow code to use new MCWeighter

2014-02-19 Steve
	* Split up the functionality for weighting the MC
		* Temporary change until everything is moved into MCWeighter
		* Added getSumw, getXsecTimesEff, getPileupWeight to SusyNtTools.
		  getEventWeight calls each of these
		* Can now request pileup systematic via sys flag
	* Now allow for NULL Met pointer in Meff methods
		* For calculating Meff/HT without met
	* Added pileup sys enums to MCWeighter: Sys_PILEUP_UP, Sys_PILEUP_DN

2014-02-12 Steve
	* Added call to clearObjects at beginning of SusyNtAna selectObjects
	* Removed LAr veto from Susy3LepCutflow

2014-02-11 Steve
	* Bugfix for muon systematics in n0150
		* Added temporary boolean argument to 
		  SusyNtAna.selectObjects: n0150BugFix
		* This flag propagates down through SusyNtTools 
		  getBaselineObjects, getPreMuons, and finally
		  Muon.setState

2014-02-10 Steve
	* Updated SusyNtTools.eventHasSusyPropagators to return
	  true as soon as a problematic sparticle is found
	* Added susySpartId1 and susySpartId2 to Event class
		* Allows us to recalculate the proc ID on the fly later

2014-02-09 Steve
	* Updated Susy3LepCutflow code
		* Updated cleaning cuts
		* More event selection methods
		* Full MC weighting, including the latest
		  prescriptions for sumw and cross section
		* Skeleton methods for histograms

2014-02-06 Steve
	* Added getBTagSFJets2Lep to SusyNtTools which returns
	  the subset of baseline jets which are tested for the b-veto.
	  These are thus the jets to be passed to the bTagSF method.

2014-02-05 Steve
	* Added MCID and proc ID to Event.print()
	* Changed from hardcoded values to JET_JVF_PT and JET_JVF_ETA in isSignalJet
	* Added template method, dumpObjects, for looping and dumping any 
	  objects with a print() method.

2014-01-29 Steve
	* Added bugfix to getCrossSection for Wh nohadtau signal samples
		* proc ID (susyFinalState) was not saved correctly
		* I will work on redesigning the approach upstream to
		  prevent similar future issues
	* Cleaned up mljj method
	* Added MCID and prod ID to Event.print()
	* Added Met.sumet to the copy constructor and assignment operator (was missing)
	* Added scripts/refreshSUSYToolsData.sh to update SUSYTools/data to trunk

2014-01-22 Steve
	* Fixed SusyNtTools getEventWeight to use 0 when proc ID -1 is found

2014-01-22 Anyes
	* Added functions for mljj 
	* Update 2LSSWh isolation cuts to tighter set (not backward compatible)

2014-01-22 Steve
	* Fixed SusyNtAna getEventWeight
		* Added useProcSumw argument. Order of args has changed,
		  so be careful!

2014-01-17 Steve
	* Changed default value of susyFinalState to 0 to be consistent with SUSYTools
	* Changed order of Hdecay enum in WhTruthExtractor. kUnknown is now 0 to be
	  consistent with susyFinalState and SUSYTools.

Root/Jet.cxx
#include "SusyNtuple/Jet.h"

#include <iostream>
#include <iomanip>

using namespace std;

using Susy::Jet;
using Susy::Particle;

/*--------------------------------------------------------------------------------*/
// Copy constructor
/*--------------------------------------------------------------------------------*/
Jet::Jet(const Jet &rhs):
  Particle(rhs),
  jvf(rhs.jvf),
  jvt(rhs.jvt),
  detEta(rhs.detEta),
  emfrac(rhs.emfrac),
  truthLabel(rhs.truthLabel),
  matchTruth(rhs.matchTruth),
  nTracks(rhs.nTracks),
  bjet(rhs.bjet),
  effscalefact(rhs.effscalefact),
  mv1(rhs.mv1),
  mv2c20(rhs.mv2c20),
  sv1plusip3d(rhs.sv1plusip3d),
  isBadVeryLoose(rhs.isBadVeryLoose),
  isHotTile(rhs.isHotTile),
  bch_corr_jet(rhs.bch_corr_jet),
  bch_corr_cell(rhs.bch_corr_cell),
  isBadMediumBCH(rhs.isBadMediumBCH),
  isBadMediumBCH_up(rhs.isBadMediumBCH_up),
  isBadMediumBCH_dn(rhs.isBadMediumBCH_dn),
  isBadTightBCH(rhs.isBadTightBCH),
  jer(rhs.jer),
  groupedNP(rhs.groupedNP),
  FTSys(rhs.FTSys)
  /*
  bjes(rhs.bjes),
  effNp(rhs.effNp),
  etaInter(rhs.etaInter),
  flavor(rhs.flavor),
  pileup(rhs.pileup),
  punchThrough(rhs.punchThrough),
  singlePart(rhs.singlePart)
  // relativeNC(rhs.relativeNC),
  */
{ 
}
/*--------------------------------------------------------------------------------*/
// Assignment operator
/*--------------------------------------------------------------------------------*/
Jet& Jet::operator=(const Jet &rhs)
{
  if (this != &rhs) {
    Particle::operator=(rhs);
    jvf  = rhs.jvf; 
    jvt  = rhs.jvt; 
    detEta = rhs.detEta;
    emfrac = rhs.emfrac;
    truthLabel = rhs.truthLabel;
    matchTruth = rhs.matchTruth;
    bjet = rhs.bjet;
    effscalefact = rhs.effscalefact;
    nTracks = rhs.nTracks;
    mv1 = rhs.mv1;
    mv2c20 = rhs.mv2c20;
    isBadVeryLoose = rhs.isBadVeryLoose;
    isHotTile = rhs.isHotTile;
    bch_corr_jet = rhs.bch_corr_jet;
    bch_corr_cell = rhs.bch_corr_cell;
    isBadMediumBCH = rhs.isBadMediumBCH;
    isBadMediumBCH_up = rhs.isBadMediumBCH_up;
    isBadMediumBCH_dn = rhs.isBadMediumBCH_dn;
    isBadTightBCH = rhs.isBadTightBCH;
    jer = rhs.jer;
    groupedNP = rhs.groupedNP;
    FTSys = rhs.FTSys;
    /*
    bjes=rhs.bjes;
    effNp=rhs.effNp;
    etaInter=rhs.etaInter;
    flavor=rhs.flavor;
    pileup=rhs.pileup;
    punchThrough=rhs.punchThrough;
    singlePart=rhs.singlePart;
    //relativeNC=rhs.relativeNC;
    */
  }
  return *this;
}
/*--------------------------------------------------------------------------------*/
// Jet Set State
/*--------------------------------------------------------------------------------*/
void Jet::setState(int sys)
{
  resetTLV();
  if(sys == NtSys::NOM) return;
  
  float sf = 0;
  if     ( sys == NtSys::JER) sf = jer;
  
  else if( sys == NtSys::JET_GroupedNP_1_UP) sf = groupedNP[0];
  else if( sys == NtSys::JET_GroupedNP_1_DN) sf = groupedNP[1];
  else if( sys == NtSys::JET_GroupedNP_2_UP) sf = groupedNP[2];
  else if( sys == NtSys::JET_GroupedNP_2_DN) sf = groupedNP[3];
  else if( sys == NtSys::JET_GroupedNP_3_UP) sf = groupedNP[4];
  else if( sys == NtSys::JET_GroupedNP_3_DN) sf = groupedNP[5];
/*
  else if( sys == NtSys::JET_BJES_Response_DN) sf = bjes[0];
  else if( sys == NtSys::JET_BJES_Response_UP) sf = bjes[1];
  else if( sys == NtSys::JET_EffectiveNP_1_DN) sf = effNp[0];
  else if( sys == NtSys::JET_EffectiveNP_1_UP) sf = effNp[1];
  else if( sys == NtSys::JET_EffectiveNP_2_DN) sf = effNp[2];
  else if( sys == NtSys::JET_EffectiveNP_2_UP) sf = effNp[3];
  else if( sys == NtSys::JET_EffectiveNP_3_DN) sf = effNp[4];
  else if( sys == NtSys::JET_EffectiveNP_3_UP) sf = effNp[5];
  else if( sys == NtSys::JET_EffectiveNP_4_DN) sf = effNp[6];
  else if( sys == NtSys::JET_EffectiveNP_4_UP) sf = effNp[7];
  else if( sys == NtSys::JET_EffectiveNP_5_DN) sf = effNp[8];
  else if( sys == NtSys::JET_EffectiveNP_5_UP) sf = effNp[9];
  else if( sys == NtSys::JET_EffectiveNP_6restTerm_DN) sf = effNp[10];
  else if( sys == NtSys::JET_EffectiveNP_6restTerm_UP) sf =  effNp[11];
  else if( sys == NtSys::JET_EtaIntercalibration_Modelling_DN) sf = etaInter[0];
  else if( sys == NtSys::JET_EtaIntercalibration_Modelling_UP) sf = etaInter[1];
  else if( sys == NtSys::JET_EtaIntercalibration_TotalStat_DN) sf = etaInter[2];
  else if( sys == NtSys::JET_EtaIntercalibration_TotalStat_UP) sf = etaInter[3];
  else if( sys == NtSys::JET_Flavor_Composition_DN) sf = flavor[0];
  else if( sys == NtSys::JET_Flavor_Composition_UP) sf = flavor[1];
  else if( sys == NtSys::JET_Flavor_Response_DN) sf = flavor[2];
  else if( sys == NtSys::JET_Flavor_Response_UP) sf = flavor[3];
  else if( sys == NtSys::JET_Pileup_OffsetMu_DN) sf = pileup[0];
  else if( sys == NtSys::JET_Pileup_OffsetMu_UP) sf =  pileup[1];
  else if( sys == NtSys::JET_Pileup_OffsetNPV_DN) sf = pileup[2];
  else if( sys == NtSys::JET_Pileup_OffsetNPV_UP) sf = pileup[3];
  else if( sys == NtSys::JET_Pileup_PtTerm_DN) sf =  pileup[4];
  else if( sys == NtSys::JET_Pileup_PtTerm_UP) sf =  pileup[5];
  else if( sys == NtSys::JET_Pileup_RhoTopology_DN) sf =  pileup[6];
  else if( sys == NtSys::JET_Pileup_RhoTopology_UP) sf =  pileup[7];
  else if( sys == NtSys::JET_PunchThrough_MC12_DN) sf = punchThrough[0];
  else if( sys == NtSys::JET_PunchThrough_MC12_UP) sf = punchThrough[1];
  //else if( sys == NtSys::JET_RelativeNonClosure_MC12_DN) sf = relativeNC[0];
  //else if( sys == NtSys::JET_RelativeNonClosure_MC12_UP) sf = relativeNC[1];
  else if( sys == NtSys::JET_SingleParticle_HighPt_DN) sf = singlePart[0];
  else if( sys == NtSys::JET_SingleParticle_HighPt_UP) sf = singlePart[1];
*/
  else return;

  this->SetPtEtaPhiE(sf * this->Pt(), this->Eta(), this->Phi(), sf * this->E());
}
/*--------------------------------------------------------------------------------*/
// Flavor systematics
/*--------------------------------------------------------------------------------*/
float Jet::getFTSys(Susy::NtSys::SusyNtSys sys)
{
    float s= 1;
    
    if      ( sys == NtSys::FT_Eigen_B_0_DN) s = FTSys[0];
    else if ( sys == NtSys::FT_Eigen_B_0_UP) s = FTSys[1];
    else if ( sys == NtSys::FT_Eigen_B_1_DN) s = FTSys[2];
    else if ( sys == NtSys::FT_Eigen_B_1_UP) s = FTSys[3];
    else if ( sys == NtSys::FT_Eigen_B_2_DN) s = FTSys[4];
    else if ( sys == NtSys::FT_Eigen_B_2_UP) s = FTSys[5];
    else if ( sys == NtSys::FT_Eigen_B_3_DN) s = FTSys[6];
    else if ( sys == NtSys::FT_Eigen_B_3_UP) s = FTSys[7];
    else if ( sys == NtSys::FT_Eigen_B_4_DN) s = FTSys[8];
    else if ( sys == NtSys::FT_Eigen_B_4_UP) s = FTSys[9];
    else if ( sys == NtSys::FT_Eigen_B_5_DN) s = FTSys[10];
    else if ( sys == NtSys::FT_Eigen_B_5_UP) s = FTSys[11];
    else if ( sys == NtSys::FT_Eigen_B_6_DN) s = FTSys[12];
    else if ( sys == NtSys::FT_Eigen_B_6_UP) s = FTSys[13];
    else if ( sys == NtSys::FT_Eigen_B_7_DN) s = FTSys[14];   
    else if ( sys == NtSys::FT_Eigen_B_7_UP) s = FTSys[15];
    else if ( sys == NtSys::FT_Eigen_B_8_DN) s = FTSys[16];
    else if ( sys == NtSys::FT_Eigen_B_8_UP) s = FTSys[17];
    else if ( sys == NtSys::FT_Eigen_B_9_DN) s = FTSys[18];
    else if ( sys == NtSys::FT_Eigen_B_9_UP) s = FTSys[19];

    else if ( sys == NtSys::FT_Eigen_C_0_DN) s = FTSys[20];
    else if ( sys == NtSys::FT_Eigen_C_0_UP) s = FTSys[21];
    else if ( sys == NtSys::FT_Eigen_C_1_DN) s = FTSys[22];
    else if ( sys == NtSys::FT_Eigen_C_1_UP) s = FTSys[23];
    else if ( sys == NtSys::FT_Eigen_C_2_DN) s = FTSys[24];
    else if ( sys == NtSys::FT_Eigen_C_2_UP) s = FTSys[25];
    else if ( sys == NtSys::FT_Eigen_C_3_DN) s = FTSys[26];
    else if ( sys == NtSys::FT_Eigen_C_3_UP) s = FTSys[27];

    else if ( sys == NtSys::FT_Eigen_Light_0_DN) s = FTSys[28];
    else if ( sys == NtSys::FT_Eigen_Light_0_UP) s = FTSys[29];
    else if ( sys == NtSys::FT_Eigen_Light_1_DN) s = FTSys[30];
    else if ( sys == NtSys::FT_Eigen_Light_1_UP) s = FTSys[31];
    else if ( sys == NtSys::FT_Eigen_Light_2_DN) s = FTSys[32];
    else if ( sys == NtSys::FT_Eigen_Light_2_UP) s = FTSys[33];
    else if ( sys == NtSys::FT_Eigen_Light_3_DN) s = FTSys[34];
    else if ( sys == NtSys::FT_Eigen_Light_3_UP) s = FTSys[35];
    else if ( sys == NtSys::FT_Eigen_Light_4_DN) s = FTSys[36];
    else if ( sys == NtSys::FT_Eigen_Light_4_UP) s = FTSys[37];
    else if ( sys == NtSys::FT_Eigen_Light_5_DN) s = FTSys[38];
    else if ( sys == NtSys::FT_Eigen_Light_5_UP) s = FTSys[39];
    else if ( sys == NtSys::FT_Eigen_Light_6_DN) s = FTSys[40];
    else if ( sys == NtSys::FT_Eigen_Light_6_UP) s = FTSys[41];
    else if ( sys == NtSys::FT_Eigen_Light_7_DN) s = FTSys[42];
    else if ( sys == NtSys::FT_Eigen_Light_7_UP) s = FTSys[43];
    else if ( sys == NtSys::FT_Eigen_Light_8_DN) s = FTSys[44];
    else if ( sys == NtSys::FT_Eigen_Light_8_UP) s = FTSys[45];
    else if ( sys == NtSys::FT_Eigen_Light_9_DN) s = FTSys[46];
    else if ( sys == NtSys::FT_Eigen_Light_9_UP) s = FTSys[47];
    else if ( sys == NtSys::FT_Eigen_Light_10_DN) s = FTSys[48];
    else if ( sys == NtSys::FT_Eigen_Light_10_UP) s = FTSys[49];
    else if ( sys == NtSys::FT_Eigen_Light_11_DN) s = FTSys[50];
    else if ( sys == NtSys::FT_Eigen_Light_11_UP) s = FTSys[51];

    return s;

}

void Jet::setFTSys(Susy::NtSys::SusyNtSys sys, double scale=1.)
{
    if      ( sys == NtSys::FT_Eigen_B_0_DN) FTSys[0] = scale;
    else if ( sys == NtSys::FT_Eigen_B_0_UP) FTSys[1] = scale;
    else if ( sys == NtSys::FT_Eigen_B_1_DN) FTSys[2] = scale;
    else if ( sys == NtSys::FT_Eigen_B_1_UP) FTSys[3] = scale;
    else if ( sys == NtSys::FT_Eigen_B_2_DN) FTSys[4] = scale;
    else if ( sys == NtSys::FT_Eigen_B_2_UP) FTSys[5] = scale;
    else if ( sys == NtSys::FT_Eigen_B_3_DN) FTSys[6] = scale;
    else if ( sys == NtSys::FT_Eigen_B_3_UP) FTSys[7] = scale;
    else if ( sys == NtSys::FT_Eigen_B_4_DN) FTSys[8] = scale;
    else if ( sys == NtSys::FT_Eigen_B_4_UP) FTSys[9] = scale;
    else if ( sys == NtSys::FT_Eigen_B_5_DN) FTSys[10] = scale;
    else if ( sys == NtSys::FT_Eigen_B_5_UP) FTSys[11] = scale;
    else if ( sys == NtSys::FT_Eigen_B_6_DN) FTSys[12] = scale;
    else if ( sys == NtSys::FT_Eigen_B_6_UP) FTSys[13] = scale;
    else if ( sys == NtSys::FT_Eigen_B_7_DN) FTSys[14] = scale;   
    else if ( sys == NtSys::FT_Eigen_B_7_UP) FTSys[15] = scale;
    else if ( sys == NtSys::FT_Eigen_B_8_DN) FTSys[16] = scale;
    else if ( sys == NtSys::FT_Eigen_B_8_UP) FTSys[17] = scale;
    else if ( sys == NtSys::FT_Eigen_B_9_DN) FTSys[18] = scale;
    else if ( sys == NtSys::FT_Eigen_B_9_UP) FTSys[19] = scale;

    else if ( sys == NtSys::FT_Eigen_C_0_DN) FTSys[20] = scale;
    else if ( sys == NtSys::FT_Eigen_C_0_UP) FTSys[21] = scale;
    else if ( sys == NtSys::FT_Eigen_C_1_DN) FTSys[22] = scale;
    else if ( sys == NtSys::FT_Eigen_C_1_UP) FTSys[23] = scale;
    else if ( sys == NtSys::FT_Eigen_C_2_DN) FTSys[24] = scale;
    else if ( sys == NtSys::FT_Eigen_C_2_UP) FTSys[25] = scale;
    else if ( sys == NtSys::FT_Eigen_C_3_DN) FTSys[26] = scale;
    else if ( sys == NtSys::FT_Eigen_C_3_UP) FTSys[27] = scale;

    else if ( sys == NtSys::FT_Eigen_Light_0_DN) FTSys[28] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_0_UP) FTSys[29] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_1_DN) FTSys[30] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_1_UP) FTSys[31] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_2_DN) FTSys[32] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_2_UP) FTSys[33] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_3_DN) FTSys[34] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_3_UP) FTSys[35] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_4_DN) FTSys[36] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_4_UP) FTSys[37] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_5_DN) FTSys[38] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_5_UP) FTSys[39] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_6_DN) FTSys[40] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_6_UP) FTSys[41] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_7_DN) FTSys[42] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_7_UP) FTSys[43] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_8_DN) FTSys[44] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_8_UP) FTSys[45] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_9_DN) FTSys[46] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_9_UP) FTSys[47] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_10_DN) FTSys[48] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_10_UP) FTSys[49] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_11_DN) FTSys[50] = scale;
    else if ( sys == NtSys::FT_Eigen_Light_11_UP) FTSys[51] = scale;

    return;
}


/*--------------------------------------------------------------------------------*/
// Jet print
/*--------------------------------------------------------------------------------*/
void Jet::print() const
{
  cout.precision(2);
  cout << fixed << "Jet : pt " << setw(6) << Pt() << " eta " << setw(5) << Eta()
       << " phi " << setw(5) << Phi()
       << " jvf " << setw(4) << jvf
       << " jvt " << setw(4) << jvt
       << " mv1 " << setw(5) << mv1
       // No way do we want to see this every time we print
       //<< " jer " << jer << " jes_up " << jes_up << " jes_dn " << jes_dn
       << endl;
  cout.precision(6);
  cout.unsetf(ios_base::fixed);
}

